near-contract-standards/src/non_fungible_token/payout/mod.rs

```diff
@@ -0,0 +1,112 @@
+use near_sdk::borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::TreeMap;
+use near_sdk::json_types::U128;
+use near_sdk::serde::{Deserialize, Serialize};
+use near_sdk::AccountId;
+use std::collections::HashMap;
+
+type BasisPoint = u16;
+
+/// This struct represents NFT royalty payout for each receiver.
+#[derive(Default, BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+#[borsh(crate = "near_sdk::borsh")]
+pub struct Payout {
+    pub payout: HashMap<AccountId, U128>,
+}
+
+/// This struct represents percentage of total royalty per receiver as well as the total percentage
+/// of distributed royalties based on incoming payment.
+#[derive(BorshDeserialize, BorshSerialize, Debug)]
+#[borsh(crate = "near_sdk::borsh")]
+pub struct Royalties {
+    /// A mapping of accounts to the percentage of total royalty to be distributed per token.
+    pub accounts: TreeMap<AccountId, HashMap<TokenId, BasisPoint>>,
+}
+
+/// An interface allowing non-fungible token contracts to request that financial contracts pay-out
+/// multiple receivers, enabling flexible royalty implementations.
+///
+/// [Royalties and Payouts standard]: <https://nomicon.io/Standards/Tokens/NonFungibleToken/Payout>
+///
+/// # Examples
+///
+/// ```
+/// use near_sdk::borsh::{BorshDeserialize, BorshSerialize};
+/// use near_sdk::{PanicOnDefault, AccountId, PromiseOrValue, near_bindgen, assert_one_yocto};
+/// use near_contract_standards::non_fungible_token::{core::NonFungibleTokenCore, NonFungibleToken, NonFungibleTokenPayout, payout::Payout, TokenId, Token};
+/// use near_sdk::json_types::U128;
+///
+/// #[near_bindgen]
+/// #[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+/// #[borsh(crate = "near_sdk::borsh")]
+/// pub struct Contract {
+///    tokens: NonFungibleToken,
+///}
+/// #[near_bindgen]
+/// impl NonFungibleTokenCore for Contract {
+///     #[payable]
+///    fn nft_transfer(&mut self, receiver_id: AccountId, token_id: TokenId, approval_id: Option<u64>, memo: Option<String>) {
+///        self.tokens.nft_transfer(receiver_id, token_id, approval_id, memo);
+///    }
+///
+///    #[payable]
+///    fn nft_transfer_call(&mut self, receiver_id: AccountId, token_id: TokenId, approval_id: Option<u64>, memo: Option<String>, msg: String) -> PromiseOrValue<bool> {
+///        self.tokens.nft_transfer_call(receiver_id, token_id, approval_id, memo, msg)
+///    }
+///
+///    fn nft_token(&self, token_id: TokenId) -> Option<Token> {
+///        self.tokens.nft_token(token_id)
+///    }
+///}
+/// #[near_bindgen]
+/// impl NonFungibleTokenPayout for Contract {
+///     #[allow(unused_variables)]
+///     fn nft_payout(
+///         &self,
+///         token_id: String,
+///         balance: U128,
+///         max_len_payout: Option<u32>,
+///     ) -> Payout {
+///         self.tokens.nft_payout(token_id, balance, max_len_payout)
+///     }
+///     #[payable]
+///     fn nft_transfer_payout(
```

Comment:
I realized that most of the examples lack at least demo methods on how to mint tokens and, in this case, how to assign royalties to the tokens. This would highlight the problem that in order to assign royalties one would need to access internal fields deep down and manipulate with them manually:

```rust
self.tokens.royalties.unwrap().royalties.entry(token_id).insert(royalties_to_account_id, 5)
```

Without this example this implementation is most probably won't be used and we may even argue that we may not need to merge it in the first place.

Reply:
Yes, I have actually been thinking about that too when I first glanced at the original PR, but then it got forgotten as I moved forward.
We do indeed have to expose an interface to assign royalties.

Reply:
Just going to leave this here: what we need to do is introduce a `create_royalties` method or similar that will only be accessible to nft owner.